Add: EventQueueSize, BatchSize, and TimerWakeUpSeconds config option

keyur9 opened the pull request:
## Description
Add: EventQueueSize, BatchSize, and TimerWakeUpSeconds config option
Add: Support to fetch userId from authorization header
Bump moesifapi-go dependency to version v1.0.6

## Related Issue

## Motivation and Context

## How This Has Been Tested
Setuped Tyk gateway locally, make the moesif pump, build the tyk-pump locally and test moesif pump changes by making api calls via tyk gateway which are captured by Moesif.

## Screenshots (if appropriate)

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist
- [x] Make sure you are requesting to **pull a topic/feature/bugfix branch** (right side). If pulling from your own
      fork, don't request your `master`!
- [x] Make sure you are making a pull request against the **`master` branch** (left side). Also, you should start
      *your branch* off *our latest `master`*.
- [x] My change requires a change to the documentation.
  - [ ] If you've changed APIs, describe what needs to be updated in the documentation.
- [x] I have updated the documentation accordingly.
- [ ] Modules and vendor dependencies have been updated; run `go mod tidy && go mod vendor`
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [x] Check your code additions will not fail linting checks:
  - [x] `go fmt -s`
  - [x] `go vet`

tbuchaillot commented:
@keyur9 what do you think if we add an `enable_bulk` and `bulk_config` config options for this? In that way, it can still have backward compatibility.

For example:
```
...
"enable_bulk":true,
"bulk_config":{
   "event_queue_size":1000,
   "batch_size":50,
   "timer_wake_up_seconds":2
}
...
```

keyur9 commented:
Thanks @tbuchaillot  I've added `enable_bulk` and `bulk_config` config options, and updated the PR.

tbuchaillot commented on pumps/moesif.go:
```diff
@@ -209,7 +211,35 @@ func (p *MoesifPump) Init(config interface{}) error {
+		}
+	}
+
+	api := moesifapi.NewAPI(p.moesifConf.ApplicationID, &apiEndpoint, eventQueueSize, batchSize, timerWakeupSeconds)
```
@keyur9 did the `moesifapi.NewAPI` change? 
How were the queueSize, batchSize, timerWakeupSeconds values handled before?

As I understand, if we don't provide any of those values, they're going to default and still modify the old behavior.

keyur9 replied:
@tbuchaillot Yes, `moesifapi.NewAPI` did changed. We added `queueSize, batchSize, timerWakeupSeconds` config options in the `moesifapi-go` latest version (1.0.6). If the values are not provided, we'll set the default values here - https://github.com/Moesif/moesifapi-go/blob/master/api.go#L117

tbuchaillot replied:
@keyur9 so in this case, the `enable_bulk` config option wouldn't work since it's always going to bulk right? Since it's always going to default if not values are given.

dkm199 replied:
>>As I understand, if we don't provide any of those values, they're going to default and still modify the old behavior.

@tbuchaillot main changes here is, we wanted to have to few parameters configurable so these parameters can be tweaked depending the customer environment and their load patterns etc. So As Keyur mentioned, yes we did change the moesifapi.NewAPI to take new parameters as input and these parameters have default values(equivalent to old behavior), if no bulk config is mentioned. So if we do wanna overwrite them at specific customer, we ll ask the customer to set enable_bulk and bulk_config accordingly. Hope this answers your concern, if not let me know. Thanks

tbuchaillot replied:
It makes sense. Thanks for clarifying guys.

tbuchaillot reviewed (approved):
LGTM

sedkis commented:
@matiasinsaurralde

keyur9 commented:
Hello @tbuchaillot, wanted to let you know that we've added the support to fetch user id from authorization header and updated the PR. Would appreciate if it could be reviewed before merging. Thank you.

tbuchaillot reviewed (changes requested):
Hi @keyur9 @dkm199  ! it seems it needs some changes to work.

Also, could you give us an example of a configuration + request for `authorization_header_name` and `authorization_user_id_field` ?  We were testing it to include it in our next pump release but we're not sure how it works.

dgilling commented on pumps/moesif.go:
```diff
@@ -209,7 +232,35 @@ func (p *MoesifPump) Init(config interface{}) error {
+		}
+
+		// Try to fetch the timer wake up seconds from the bulk config
+		if timer, found := p.moesifConf.BulkConfig["timer_wake_up_seconds"].(int); found {
```
Hi @tbuchaillot, by default Moesif maps a Tyk Token alias to a Moesif user id. While this is fine for most Tyk customers, some have requested the ability to read the subject within a JWT itself such as if using OpenId connect. As example, one would set `authorization_header_name` to `Authorization` and `authorization_user_id_field` to `sub` (for subject). You can test this by adding a JWT to the Authorization header such as from https://jwt.io/

konrad-sol commented:
thanks @keyur9 it works fine now

tbuchaillot commented on README.md:
```diff
@@ -362,6 +362,13 @@ The Tyk Dashboard uses the "mongo-pump-aggregate" collection to display analytic
+- `"authorization_header_name"` - (optional) An optional request header field name to used to identify the User in Moesif. Type: String. Default value is `authorization`.
+- `"authorization_user_id_field"` - (optional) An optional field name use to parse the User from authorization header in Moesif. Type: String. Default value is `sub`.
+- `"enable_bulk"` - Set this to `true` to enable `bulk_config`.
+- `"bulk_config"`- (optional) Batch writing trigger configuration.
```
Make sense, thanks for clarifying !

sedkis commented:
@matiasinsaurralde @tbuchaillot can we merge if approved?

tbuchaillot commented:
merged! 

Thanks for the contribution @keyur9 @dkm199 !
